File: 1043-partition-array-for-maximum-sum/1043-partition-array-for-maximum-sum.java
class Solution {
    private int solveRec(int []arr, int i , int k){
        if(i==arr.length) return 0;
        int len=0, maxi=Integer.MIN_VALUE, maxAns=Integer.MIN_VALUE;
        for(int j=i; j<Math.min(arr.length, i+k); j++){
            len++;
            maxi= Math.max(maxi, arr[j]);
            int sum = (len*maxi) + solveRec(arr, j+1, k);
            maxAns = Math.max(maxAns, sum);
        }
        return maxAns;
    }
    private int solveMem(int []arr, int i , int k, Integer []dp){
        if(i==arr.length) return 0;
        if(dp[i]!=null) return dp[i];
        int len=0, maxi=Integer.MIN_VALUE, maxAns=Integer.MIN_VALUE;
        for(int j=i; j<Math.min(arr.length, i+k); j++){
            len++;
            maxi= Math.max(maxi, arr[j]);
            int sum = (len*maxi) + solveMem(arr, j+1, k, dp);
            maxAns = Math.max(maxAns, sum);
        }
        return dp[i] = maxAns;
    }
    public int maxSumAfterPartitioning(int[] arr, int k) {
        // // Approach-1: Recursion 
        // return solveRec(arr, 0, k);
        
        // Approach-2: Recursion + Mamoization 
        Integer []dp =new Integer[arr.length];
        return solveMem(arr, 0, k, dp);
    }
}

File: 1043-partition-array-for-maximum-sum/1043-partition-array-for-maximum-sum.cpp
class Solution {
public:
    int solveMem(vector<int>& arr, int i, int k, vector<int>&dp){
        int n = arr.size();
        if(i==arr.size()) return 0;
        if(dp[i]!=-1) return dp[i];
        int len=0, maxi=INT_MIN, maxAns = INT_MIN;
        for(int j=i; j<min(n, i+k); j++){
            len++;
            maxi = max(maxi, arr[j]);
            int sum= len*maxi + solveMem(arr, j+1, k, dp);
            maxAns= max(maxAns, sum);
        }
        return dp[i] = maxAns;
    }
    int solveTab(vector<int>&arr, int k){
        int n =arr.size();
        vector<int>dp(n+1, 0);
       for (int i = n - 1; i >= 0; i--) {
            int len = 0, maxi = INT_MIN, maxAns = INT_MIN;

            for (int j = i; j < min(n, i + k); j++) {
                len++;
                maxi = max(maxi, arr[j]);
                int sum = len * maxi + dp[j + 1];
                maxAns = max(maxAns, sum);
            }

            dp[i] = maxAns;
        }
        return dp[0];
    }
    int maxSumAfterPartitioning(vector<int>& arr, int k) {
        // // Memoization
        // vector<int>dp(arr.size(), -1);
        // return solveMem(arr, 0, k, dp);
        
        // Tabulation 
        return solveTab(arr, k);
    }
};